Merge 'Feature manager'

Mephiles opened the pull request:
Every page using FeatureManager should run it at the start using `let featureManager = new FeatureManager();`. I didn't test it but I think it's better to isolate it to certain pages rather than running in global scope.

Instead of running page's content like this:
```
requireContent().then(async () => {
		await displayNetworth().catch((error) => console.error("Couldn't load the live networth.", error));
		await displayEffectiveBattleStats().catch((error) => console.error("Couldn't load the effective battle stats.", error));
	});
```
It will be run as (FeatureManager will log any errors so no need to catch them):
```
requireContent().then(async () => {
		featureManager.clear();
		featureManager.load({
			name: 'Live Networth',
			enabled: !!(settings.apiUsage.user.networth && settings.pages.home.networthDetails && hasAPIData()),
			func: displayNetworth,
			runWhenDisabled: true
		});
		featureManager.load({
			name: 'Effective Battle Stats',
			enabled: settings.pages.home.effectiveStats,
			func: displayEffectiveBattleStats,
			runWhenDisabled: true
		});
	});
```
This will create new items in the FeatureManager. Following is the current format how a feature is added:
```
feature = {
	name: string,
	enabled: boolean,
	func: function,
	runWhenDisabled: boolean
}
```

I tested it a bit and it runs smoothly live when changing settings on the preferences page.

Mephiles commented:
Single feature reloads will be done like this: `featureManager.reload('Live Networth');`

Mephiles commented:
I see I forgot to complete the collapsing of the container so don't merge it before I add it too.

DeKleineKobini commented:
I would do it in the global state though, as we have features that should go be shown in ttGlobal. 

It would probably also make it easier to add a setting to disable that box.

Mephiles commented:
Update:
FeatureManager is not initiated in the `ttGlobal.entry.js` file, no need to run it in every file separately anymore. I also changed `ttGlobal.entry.js, ttHome.js, ttItems.js` into OOP format, thus increasing readability.

Every method in `[Page]Features` class that is running a feature should now firstly add the feature by doing:
```
featureManager.new({
	name: "Align Left",
	scope: "global",
	enabled: settings.pages.global.alignLeft,
	func: feature,
	runWhenDisabled: true,
});
```
All features are then loaded by doing:
```
featureManager.load("Align Left");
featureManager.load("Hide Level Upgrade");
featureManager.load("Hide Leave Buttons");
featureManager.load("Hide Icons");
featureManager.load("Hide Areas");
```

I noticed that the class solution might get a bit clustered sometimes but I think it's better and still more readable than 10 different functions in the main scope. Let me know how it looks/feels, DKK.

--
Finally, added a collapsing possibility. Have one more thing to do, implement scopes as seen in the previous example. This would just categorize the features in the container so the user understands better what features are run globally and what on the current page.

DeKleineKobini commented:
I'm honestly not sure how I feel about having the features in classes. It seems unnecessary, as you are just moving the functions down 1 level, it's not like they are getting split into logical parts.

The container looks great, except when closed. The arrow seems to be at a strange position when closed.

I also think there should be some extra settings:
- [ ] Don't show disabled features.
- [ ] Only show failed features.
- [ ] Display container in the left space.

DeKleineKobini commented on extension/scripts/content/global/ttGlobal.entry.js:
```diff
@@ -1,70 +1,234 @@
+		featureManager.new({
+			name: "Chat Font Size",
+			scope: "global",
+			enabled: settings.pages.chat.fontSize,
```
I would say that this feature isn't enabled when the fontSize = 12 (Torn default).

Mephiles replied:
Changed.

DeKleineKobini commented on extension/scripts/content/global/ttGlobal.entry.js:
```diff
@@ -1,70 +1,234 @@
+			runWhenDisabled: true,
+		});
+
+		async function feature_1() {
```
Not a fan of these function names.

Mephiles replied:
Changed.

DeKleineKobini commented on extension/scripts/content/global/ttGlobal.entry.js:
```diff
@@ -1,70 +1,234 @@
+	globalFeatures.init();
 
-	document.documentElement.style.setProperty("--torntools-hide-upgrade-button", settings.pages.global.hideLevelUpgrade ? "none" : "block");
+	storageListeners.settings.push(function () {
```
To be consistent, it should be an arrow function.

DeKleineKobini replied:
This also is the case on other places.

Mephiles replied:
Changed in all places I noticed :P

DeKleineKobini commented on extension/scripts/content/global/ttGlobal.entry.js:
```diff
@@ -1,70 +1,234 @@
+		this.chatFontSize();
 
-	document.documentElement.style.setProperty("--torntools-chat-font-size", `${settings.pages.chat.fontSize || 12}px`);
+		if (hasAPIData()) this.highlightRefills();
```
Should we do these kind of checks here or in the function itself?

Mephiles replied:
Moved them to the function. Will have to test functionality just in case

DeKleineKobini commented on extension/scripts/global/globalClasses.js:
```diff
@@ -7,3 +7,174 @@ class DefaultSetting {
+
+		document.find(".tt-page-status-header").onclick = function () {
+			this.classList.toggle("collapsed");
+			ttStorage.change({ filters: { containers: { [containerID]: this.classList.contains("collapsed") } } });
```
```
Uncaught ReferenceError: containerID is not defined
    at HTMLDivElement.document.find.onclick (globalClasses.js:70)
```
Should be `[this.containerID]` I guess.

Mephiles replied:
Fixed.

DeKleineKobini commented:
![Screenshot](https://i.imgur.com/xOj1G4U.png)

This is what I meant with strange display when closed.

DeKleineKobini commented on extension/scripts/content/global/ttGlobal.entry.js:
```diff
@@ -1,70 +1,234 @@
+		featureManager.display(settings.featureDisplay);
+	});
+	storageListeners.userdata.push(function () {
+		globalFeatures.init();
```
This reloads more features than it should.

Mephiles replied:
It runs the same way as previously.

DeKleineKobini replied:
Seems it does, but it should be changed I guess, especially since they are now separated.

DeKleineKobini commented on extension/scripts/content/home/ttHome.js:
```diff
@@ -8,202 +8,254 @@ let networthInterval = false;
+
+	storageListeners.settings.push(function () {
+		homeFeatures.init();
+		featureManager.display(settings.featureDisplay);
```
Isn't this handled in the ttGlobal.entry already?

Mephiles replied:
Yep, my bad. Fixed.

DeKleineKobini commented on extension/scripts/content/items/ttItems.js:
```diff
@@ -6,552 +6,608 @@ let pendingActions = {};
-	loadItemsOnce();
+	storageListeners.settings.push(function () {
+		itemsFeatures.init();
+		featureManager.display(settings.featureDisplay);
```
Isn't this handled in the ttGlobal.entry already?

Mephiles replied:
Yep, my bad. Fixed.

DeKleineKobini commented:
I would also place the container a bit more to the right.

125% zoom (windows recommended for my laptop)
![125% zoom](https://i.imgur.com/Un1g4NQ.png)

150% zoom (what I actually use on my laptop), I do know that this is not really fixable
![150% zoom](https://i.imgur.com/bv3HDyk.png)

Mephiles commented:
I allowed the container in all 4 corners (user's choice). Top corners are absolute elements (they disappear when you scroll down). Bottom corners are fixed elements. They are in relation to the window border not Torn's content.

DeKleineKobini commented on extension/scripts/global/globalClasses.js:
```diff
@@ -38,10 +37,11 @@ class FeatureManager {
 		let collapsed = this.containerID in filters.containers ? filters.containers[this.containerID] : false;
 
-		document.find(".content").appendChild(
+		document.find("body").appendChild(
```
`document.body` would be better, in this case a find is not needed

Mephiles replied:
I always forget that :D

DeKleineKobini commented on extension/scripts/global/globalClasses.js:
```diff
@@ -67,7 +67,7 @@ class FeatureManager {
 		document.find(".tt-page-status-header").onclick = function () {
 			this.classList.toggle("collapsed");
-			ttStorage.change({ filters: { containers: { [containerID]: this.classList.contains("collapsed") } } });
+			ttStorage.change({ filters: { containers: { [this.containerID]: this.classList.contains("collapsed") } } });
```
Did you test this? Wouldn't `this` go to the element instead of the class?

Mephiles replied:
Just pushed a fix for this.

DeKleineKobini commented on extension/scripts/global/globalClasses.js:
```diff
@@ -65,9 +65,9 @@ class FeatureManager {
-			ttStorage.change({ filters: { containers: { [this.containerID]: this.classList.contains("collapsed") } } });
+		document.find(".tt-page-status-header").onclick = () => {
+			document.find(".tt-page-status-header").classList.toggle("collapsed");
+			ttStorage.change({ filters: { containers: { [this.containerID]: document.find(".tt-page-status-header").classList.contains("collapsed") } } });
```
`classList.toggle` returns a boolean, to show whether or nor is now has the class or not, you can use this for the collapsed or not, instead of having to find the element again.

Mephiles replied:
Oh, thanks! Wasn't aware it returned anything.

DeKleineKobini replied:
I didn't even know it existed at first xD

DeKleineKobini commented:
Closing this as this has been 'merged' manually.
